web/html/src/manager/proxy/container-config.tsx

```diff
@@ -0,0 +1,359 @@
+            const reader = new FileReader();
+            reader.onload = (e) => {
+              if (e.target?.result instanceof ArrayBuffer) {
+                // Should never happen since we call readAsText, just quiets tsc
+                resolve(undefined);
+              } else {
+                resolve({ [fieldName]: replaceCRLF(e.target?.result) });
+              }
+            };
+            reader.readAsText(file);
+          });
+        }
+        return undefined;
+      })
+      .filter((promise) => promise != null);
+    Promise.all(fileReaders).then((values) => {
+      const commonData = {
+        proxyFQDN: model.proxyFQDN,
+        proxyPort: model.proxyPort ? parseInt(model.proxyPort, 10) : 8022,
+        serverFQDN: model.serverFQDN,
+        maxSquidCacheSize: parseInt(model.maxSquidCacheSize, 10),
+        proxyAdminEmail: model.proxyAdminEmail,
+        sslMode: model.sslMode,
+      };
+
+      const cnamesData = Object.fromEntries(Object.entries(model).filter(([key]) => key.startsWith("cnames")));
+      const extraData =
+        model.sslMode === SSLMode.CreateSSL
+          ? Object.assign(
+              {},
+              {
+                caPassword: model.caPassword,
+                country: model.country,
+                state: model.state,
+                city: model.city,
+                org: model.org,
+                orgUnit: model.orgUnit,
+                sslEmail: model.sslEmail,
+              },
+              cnamesData
+            )
+          : {};
+      const formData = unflattenModel(Object.assign({}, commonData, extraData, ...values));
+      Network.post("/rhn/manager/api/proxy/container-config", formData).then(
+        (data) => {
+          setSuccess(data.success);
+          setMessages([]);
+          setLoading(false);
+          window.location.assign("/rhn/manager/api/proxy/container-config/" + data);
+        },
+        (xhr) => {
+          try {
+            setSuccess(false);
+            setMessages([JSON.parse(xhr.responseText)]);
+            setLoading(false);
+          } catch (err) {
+            const errMessages = DEPRECATED_unsafeEquals(xhr.status, 0)
+              ? t("Request interrupted or invalid response received from the server.")
+              : Network.errorMessageByStatus(xhr.status)[0];
+            setSuccess(false);
+            setMessages([errMessages]);
+            setLoading(false);
+          }
+        }
+      );
+    });
+  };
+
+  const clearFields = () => {
+    setModel(initialModel);
+  };
+
+  const onValidate = (isValid: boolean) => {
+    setInvalid(!isValid);
+  };
+
+  const onChange = (newModel: any) => {
+    setModel(Object.assign({}, newModel));
+  };
+
+  const onAddField = (fieldName: string) => {
+    return (index: number) => setModel(Object.assign({}, model, { [fieldName + index]: "" }));
+  };
+
+  const onRemoveField = (fieldName: string) => {
+    return (index: number) => {
+      var newModel = model;
```

Comment:
This variable seems unused? There should be no difference whether you `delete` from this object or the other one since they point to the same reference unless I'm missing something?

Reply:
If we don't do the `setModel(newModel)` afterwards, the state doesn't get updated... and the form is not rendered again

Reply:
My point is there is no "new" model here, the variables refer to the exact same object. The code here is in essence:  

```ts
const a = { foo: 1, bar: 2 };
const b = a;
delete b["foo"];
a === b; // true
a.foo === undefined; // true
b.foo === undefined; // true
```

If you want to extract the value into a new reference, the easiest way would be a-la shallow copy:  

```ts
const newModel = { ...model };
delete newModel[...];
setModel(newModel);
```

Reply:
indeed `new` doesn't really describe  the mode, I should say `updatedModel`. I still don't get why we need a shallow copy to do this...

Reply:
Javascript objects aren't passed by value, they're passed by reference. To exemplify with imaginary memory addresses:

```ts
const a = {}; // Create a new object and make `a` point to it, `a` points to 0x00
const b = {}; // Create a new object and make `b` point to it, `b` points to 0x01
const c = b; // Make `c` point to the same memory address as `b`, `c` also points to 0x01, it is NOT a copy of the object stored at 0x01
```

This isn't always clear unless you keep it in mind while reading code, but has implications, such as:  

```ts
b.foo = "foo"; // Add the property `foo` to the object stored at 0x01
c.foo === "foo"; // True even though we didn't touch `c`, because `c` points to the same memory address as `b`
```

A shallow copy is a way to cheaply get a new reference, you can think of it as two steps:

```ts
const d = { // Create a new object and make `d` point to it, `d` points to 0x02
  ...b, // Shallow copy all values from the object stored at 0x01 into this object
};
d.bar = "bar"; // Add the property `bar` to the object stored at 0x02
b.bar !== "bar"; // True because `d` and `b` point to different memory addresses, the `bar` property was added only to the one at 0x02, but not the one at 0x01
```

Hope this helps make it a bit clearer.